File: backend/supabase/shineservice/endpoint/toggle-ray.go
package SupabaseShinesEndpoint

import (
	"encoding/json"
	"log"
	"net/http"
	SupabaseShines "services/supabase/shineservice"

	sharedCtx "services/utils/context"

	"github.com/go-chi/chi/v5"
)

func ToggleRayHandler(shineSvc *SupabaseShines.Service) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			http.Error(w, "Method Not Allowed", http.StatusMethodNotAllowed)
			return
		}

		shineId := chi.URLParam(r, "shineId")
		if shineId == "" {
			http.Error(w, "Shine ID is required", http.StatusBadRequest)
			return
		}

		uid, ok := r.Context().Value(sharedCtx.UIDKey).(string)
		if !ok || uid == "" {
			http.Error(w, "Unauthorized", http.StatusUnauthorized)
			return
		}

		rayToggled, err := shineSvc.ToggleRay(r.Context(), uid, shineId)
		if err != nil {
			log.Printf("API Error: Failed to toggle ray for shine %s by user %s: %v", shineId, uid, err)
			if err.Error() == "shine post does not exist" {
				http.Error(w, "Shine not found", http.StatusNotFound)
				return
			}
			http.Error(w, "Failed to process request", http.StatusInternalServerError)
			return
		}

		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusCreated)
		if err := json.NewEncoder(w).Encode(rayToggled); err != nil {
			log.Printf("API Error: Failed to encode response: %v", err)
			http.Error(w, "Failed to encode response", http.StatusInternalServerError)
		}
	}
}

File: web/pages/api/shines/[shineId]/toggleRay.ts
// api/shines/[shineId]/toggleRay

import type { NextApiRequest, NextApiResponse } from "next";
import { corsInstance, runMiddleware, tokenAuthorization } from "@/lib/authorization/helper";
import { toggleRay } from "@/lib/firebase/adminUtils/shineService";

const cors = corsInstance(['POST', 'OPTIONS']);

//Should toggle ray and add user to list of users who have added a ray
export default async function handler(req: NextApiRequest, res: NextApiResponse){

    await runMiddleware(req, res, cors);
    if(req.method === 'OPTIONS'){
        return res.status(200).end();
    }

    if(req.method !== 'POST'){
        return res.status(405).json({ message: "Method Not Allowed." });
    }

    try{
        const requesterUid = await tokenAuthorization(req);
        const { shineId } = req.query;

        if(!shineId || typeof shineId !== 'string'){
            return res.status(400).json({ error: 'Shine ID is required as string' });
        };

        const rayToggled = await toggleRay(requesterUid, shineId);
        return res.status(201).json(rayToggled);

    } catch (err: any) {
        if(err.message === 'Unauthorized'){
            return res.status(401).json({ message: "Unauthorized" });
        }
        console.error(`API Error for api/shines/${req.query.shineId}/toggleRay`);
        return res.status(500).json({ error: err.message || "Failed to process request" });
    }
}